### PR TITLE
MixtralSparseMoeBlock: add gate jitter

### LINKED ISSUE #29429: Gate jitter is missing from Sparse MoE routing implementation
### System Info

N/A

### Who can help?

@ArthurZucker 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

In the official Mixtral MegaBlocks fork, there is a mention of [gate jitter.](https://github.com/mistralai/megablocks-public/blob/15634e007ecfff64cf19b4eefccdb5de4c1fe7c9/megablocks/layers/router.py#L40)

If you look at [modeling_mixtral.py](https://github.com/huggingface/transformers/blob/39ef3fb248ba288897f35337f4086054c69332e5/src/transformers/models/mixtral/modeling_mixtral.py), you can note that there is no mention of gate jitter, which serves as a regularization technique for MoE routing (I believe it is also described as "[Noisy gating](https://paul-bruffett.medium.com/mixture-of-experts-419806770255)" or in the context of [the original Sparse MoE paper](https://arxiv.org/pdf/1701.06538.pdf), as "Noisy Top-K gating".)

The jitter is implemented elsewhere in the repository for [Switch Transformers](https://github.com/huggingface/transformers/blob/704b3f74f9685e1772acd9949f65b0c5bbd64539/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L172), but it is completely absent in the sparse MoE implementation provided for Mixtral, despite it being implemented in the official MistralAI MegaBlocks fork.

### Expected behavior

I believe that `MixtralSparseMoeBlock` needs to be adjusted so that the forward method includes a call to a new jitter method (similar to the one in the LearnedRouter class of the Megablocks implementation) that applies gate jitter to the input data before passing it through the MoE layer, if it is turned on.
It is unclear whether or not the jitter was used in the first place to pretrain Mixtral, but this would bring it to feature parity.

### PR DESCRIPTION
**Pull request by lorenzoverardo**
# What does this PR do?

This PR adds gate jitter to MixtralSparseMoeBlock's input data before passing it through the MoE layer, if turned on.
Fixes https://github.com/huggingface/transformers/issues/29429.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

- text models: @ArthurZucker and @younesbelkada

### CODE REVIEW

File: src/transformers/models/mixtral/configuration_mixtral.py [hunk 1]
```diff
@@ -93,6 +93,8 @@ class MixtralConfig(PretrainedConfig):
             allow the model to output the auxiliary loss. See [here]() for more details
         router_aux_loss_coef (`float`, *optional*, defaults to 0.001):
             The aux loss factor for the total loss.
+        router_jitter_noise (`float`, *optional*, defaults to 0.01):
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/configuration_mixtral.py:96]: thanks for following the API conventions! Let's default to 0 to make sur this is BC

File: src/transformers/models/mixtral/configuration_mixtral.py [hunk 2]
```diff
@@ -134,6 +136,7 @@ def __init__(
         num_local_experts=8,
         output_router_logits=False,
         router_aux_loss_coef=0.001,
+        router_jitter_noise=0.01,
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/configuration_mixtral.py:139]:
```suggestion
        router_jitter_noise=0,
```

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 3]
```diff
@@ -837,9 +837,21 @@ def __init__(self, config):
...
+        # Jitter parameters
+        self.jitter_noise = config.router_jitter_noise
+
+    def jitter(self, x):
+        if self.training and self.jitter_noise > 0:
+            low = 1.0 - self.jitter_noise
+            high = 1.0 + self.jitter_noise
+            noise = torch.rand(x.size(), dtype=x.dtype, device=x.device)
+            return x * (low + noise * (high - low))
+        return x
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:849]: no need for a new function, let's do something like https://github.com/huggingface/transformers/blob/b345f4e7f7f852f7def0ebe473d21419ae235fdc/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L163

**Review by ArthurZucker — COMMENTED** [review]
Thanks!

File: src/transformers/models/mixtral/configuration_mixtral.py (hunk 1 above)
Comment (lorenzoverardo) [reply src/transformers/models/mixtral/configuration_mixtral.py]: I've defaulted it to `0.0` as it's a float, and to keep it consistent with the other args 🙂

**Review by ArthurZucker — APPROVED** [review]
Last nit and good to go!

**Comment by ArthurZucker** [comment]
Thanks for the contribution!

**Comment by ArthurZucker** [comment]
FYI @younesbelkada

**Pull request merged**